src/context.rs

```diff
@@ -556,17 +568,55 @@ pub trait SeafowlContext: Send + Sync {
 }
 
 impl DefaultSeafowlContext {
+    // Create a new `DefaultSeafowlContext` with a new inner context scoped to a different default DB
+    pub fn scope_to_database(&self, name: String) -> Result<Arc<DefaultSeafowlContext>> {
+        let database_id =
+            self.all_database_ids
+                .read()
+                .get(name.as_str())
+                .map(|db_id| *db_id as DatabaseId)
+                .ok_or_else(|| {
+                    DataFusionError::Plan(format!(
+                        "Unknown database {name}; try creating one with CREATE DATABASE first"
+                    ))
+                })?;
+
+        // Swap the default database in the new internal context's session config
+        let session_config = self
+            .inner()
+            .copied_config()
+            .with_default_catalog_and_schema(name.clone(), DEFAULT_SCHEMA);
+
+        Ok(Arc::from(DefaultSeafowlContext {
+            inner: SessionContext::with_config_rt(
+                session_config,
+                self.inner().runtime_env(),
+            ),
+            table_catalog: self.table_catalog.clone(),
```

Comment:
Does this mean we load the _entire_ world (the state of all tables and all schemas in all databases) and then scope the context down to a single DB? Would it be possible to load just the context for a single database at the beginning of handling the HTTP request (we know its name already and we wouldn't support cross-db queries).

Reply:
Actually no, we don't load the entire world. First of, reminder that `{table, partition, function}_catalog` is only a [thin client](https://github.com/splitgraph/seafowl/blob/631de991a9dc84b7497a041c4423795f0d1d12c1/src/config/context.rs#L34-L72) for fetching the corresponding metadata during planning/execution, the only thing it stores inside is a repository (with a connection pool). These are all shared via the Arc.

The actual state is kept in the inner `SessionContext` (and the `database`/`database_id`). When we perform scoping we start off from a [blank state](https://github.com/splitgraph/seafowl/blob/multi-db-impl/src/context.rs#L591-L594) (containing only the new default DB name and default schema). The state of all existing schemas/tables is then loaded as before prior to planning each query via [reload_schema](https://github.com/splitgraph/seafowl/blob/multi-db-impl/src/context.rs#L1037). Thus the only "penalty" for targeting a different DB is re-creating a new blank `SessionContext`, no actual additional DB hits required since `CREATE DATABASE` ensures we have the name/id entry in the shared `all_database_ids` map.

So we do load the context for a single database only, but we do it per-statement not per-HTTP request (this was also the case before). This means we'll perform multiple `reload_schema`'s while processing consecutive statements from a statement batch, which I think is fine for now (and can be changed later on).

Reply:
Ah, sweet! Thanks for clarifying.